test/unit/org/apache/cassandra/db/marshal/AbstractTypeTest.java

```diff
@@ -0,0 +1,439 @@
+import org.apache.cassandra.io.util.DataOutputBuffer;
+import org.apache.cassandra.io.util.File;
+import org.apache.cassandra.schema.CQLTypeParser;
+import org.apache.cassandra.schema.ColumnMetadata;
+import org.apache.cassandra.schema.Types;
+import org.apache.cassandra.transport.ProtocolVersion;
+import org.apache.cassandra.utils.AbstractTypeGenerators;
+import org.apache.cassandra.utils.ByteBufferUtil;
+import org.apache.cassandra.utils.FastByteOperations;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+import org.quicktheories.core.Gen;
+import org.reflections.Reflections;
+import org.reflections.scanners.Scanners;
+import org.reflections.util.ConfigurationBuilder;
+
+import static org.apache.cassandra.utils.AbstractTypeGenerators.extractUDTs;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.quicktheories.QuickTheory.qt;
+
+public class AbstractTypeTest
+{
+    //TODO
+    // isCompatibleWith/isValueCompatibleWith/isSerializationCompatibleWith,
+    // withUpdatedUserType/expandUserTypes/referencesDuration - types that recursive check types
+    // getMaskedValue
+
+    @Test
+    public void allTypesCovered()
+    {
+        // this test just makes sure that all types are covered and no new type is left out
+        Reflections reflections = new Reflections(new ConfigurationBuilder()
+                                                  .forPackage("org.apache.cassandra")
+                                                  .setScanners(Scanners.SubTypes)
+                                                  .setExpandSuperTypes(true)
+                                                  .setParallel(true));
+        Set<Class<? extends AbstractType>> subTypes = reflections.getSubTypesOf(AbstractType.class);
+        Set<Class<? extends AbstractType>> coverage = AbstractTypeGenerators.knownTypes();
+        StringBuilder sb = new StringBuilder();
+        for (Class<? extends AbstractType> klass : Sets.difference(subTypes, coverage))
+        {
+            if (Modifier.isAbstract(klass.getModifiers()))
+                continue;
+            if ("test".equals(new File(klass.getProtectionDomain().getCodeSource().getLocation().getPath()).name()))
+                continue;
+            String name = klass.getCanonicalName();
+            if (name == null)
+                name = klass.getName();
+            sb.append(name).append('\n');
+        }
+        if (sb.length() > 0)
+            throw new AssertionError("Uncovered types:\n" + sb);
+    }
+
+    @Test
+    public void comparableBytes()
+    {
+        qt().withShrinkCycles(0).forAll(examples(1)).checkAssert(example -> {
+            AbstractType type = example.type;
+            for (Object value : example.samples)
+            {
+                ByteBuffer bb = type.decompose(value);
+                for (ByteComparable.Version bcv : ByteComparable.Version.values())
```

Comment:
I know there are only 2 versions and I filter 2 out... but this allows for new versions to get tested by default